### DISCUSSION TITLE
Provide a fixed URL stream to play to webradio

### DISCUSSION THREAD

**Discussion by gerjaneghuizen** [music-assistant/support#2410] [Feature requests and ideas]
Hello, I'm testing the Music-Assistant and loving it so far!!!

The only part that is not present (yet) is the ability to open up a url and use a mp3 stream to listen to a player (or i'm not finding the correct documentation). 
I'm currently using Owntone for this to stream to WebRadio's i've build in a bathroom. https://owntone.github.io/owntone-server/outputs/streaming/ 

Can this be an option within Music-Assistant?

Note: Chrome-Cast groups have the option to individually change the volume!!! COOOL

**Comment by OzGav** [comment]
If it streams to a URL then they can be added manually.

**Reply by gerjaneghuizen** [reply]
It should be a fixed URL that can be reached, like opening a webradio, only then internal towards a MusicAssistant player (so the output is not towards chromecast or so but only reachable with the URL), or that each player also has a streaming URL?

**Reply by OzGav** [reply]
Yes just one URL that is fixed. Add that to MA via the icon in the top right of the radio view. It will then appear as a “Radio station”

**Reply by gerjaneghuizen** [reply]
I would think it should end up in the below list:

![image](https://github.com/music-assistant/hass-music-assistant/assets/57642280/0c66437e-43b9-48b4-bf84-5437b8f9c23b)

on each device you are then able to point a browser or dedicated application towards the the stream like in a Busch-Radio iNet radio, the bottem one is currently Owntone URL listen to what is being played:
![image](https://github.com/music-assistant/hass-music-assistant/assets/57642280/3ad5d87c-d869-4ea5-841d-7f81e0ed3c50)

**Reply by OzGav** [reply]
Ok I am not sure what you are trying to do. The owntone URL should be able to be added to MA. You can then send that stream to any players that MA supports. Your Webradios will need to support a protocol that MA can steam to such as Slimproto.

**Reply by gerjaneghuizen** [reply]
I would like that MA also has a direct URL to listen to a stream. so you can listen in on any device like you do with a webradio. The in the MA interface you select the music/source and the output is we webradio URL.

**Reply by OzGav** [reply]
MA does use streaming URLs but the architecture means there isn’t one that is always the same. If you want to send to your Webradios then you would need to change them to run slimproto (for example) and then use MA  to control what is played on them.

**Reply by gerjaneghuizen** [reply]
you have far more knowledge about the backend and technology depth that is needed, but we are on the same page now :)
My idea was to add a player provider, stream URL orso (like ChromeCast/Sonos etc), and this will then generate a single / fixed URL for the output. This way you can select it like you do with ChromeCast.

**Reply by OzGav** [reply]
Ok I have changed the title to better reflect the idea ( I think). We will see how many other people like the concept.

**Reply by mrueg** [reply]
Coming from https://github.com/orgs/music-assistant/discussions/3006 this is something that I'd be interested in using as well. A way to broadcast audio to clients that only accept an URL similar to how a web radio does it.

**Comment by mrueg** [comment]
Here's a workaround I found:

* Install https://github.com/Poeschl-HomeAssistant-Addons/mpd (this will create an mpd media_player entity)
* Enable httpd_output in the mpd addon (which allows for web streaming)
* Use the home assistant media player plugin in music assistant and select mpd as the output

**Reply by canatella** [reply]
It seems the httpd port used by the mpd addon (8000) is not mapped and so is not available from the outside network ? https://github.com/Poeschl-HomeAssistant-Addons/mpd/blob/b63026e7d535705a8d7097bcba5f35a4f3019ace/mpd/config.yaml#L26

**Reply by mrueg** [reply]
You need to map it, just click enable disabled ports.

**Reply by canatella** [reply]
Thanks, the french translation of the option, says "afficher les ports désactivé" which means "display deactivated ports" instead of map. I'll report that.

**Reply by mac-city** [reply]
> * Use the home assistant media player plugin in music assistant and select mpd as the output

I can't figure out what steps I need to take to do this.  I have MPD running in home assistant, I think I have mapped port 8000, and a steaming url I can see is running in mpd, but don't see anything in Music Assistant to select mpd as the output.  I am not able to follow that instruction.
![Screenshot 2024-10-19 at 12 56 16 PM](https://github.com/user-attachments/assets/3f6b8c5b-3f4e-4dd3-ade1-f0a42ab6af19)

**Reply by gerjaneghuizen** [reply]
Finally figured the last step out: after adding MPD to the Add-ons, you need to add it also within HomeAssistant Devices & Services and add the MPD services (i used the internal IP address). When done you can perform the last step "home assistant media player plugin in music assistant" and the player will show up in the list.

**Reply by proton999** [reply]
Can someone take me by the hand. I'm created a local MP3 stream with Icecast on a Raspberry PI, when I open the stream it's playing great.
But I want to get this done on Music Assistant which I installed on Home Assistant with the addon. I'm also installed the MPD addon and intergration. But in MA the player is always turned off if I turn it on via HA and look in MA it says that the queue is empty while I added the stream to MPD.  I'm fighting this for a couple of days, I think i'm close but I really need help with te last step because I don't see it.

**Reply by OzGav** [reply]
There is a bunch of info here https://www.music-assistant.io/faq/stream-to/#music-assistant

**Reply by proton999** [reply]
I have read that from top to bottom several times but I don't succeed. I can't see how it works, that tells more about me probably then wat is written. So if someone has tips on what I'm missing I would be gratefull.

**Reply by rrooggiieerr** [reply]
For me it's also not entirely clear how to proceed after setting up de MPD add-on and MPD integration. Which URL should I configure in my web radio (LaMetric TIME) to listen to Music Assistant?

**Reply by OzGav** [reply]
I don’t use this so can only surmise that the MPD addon should be generating an URL and this is what you should use. MA streams to the HA MPD entity which then creates the URL stream?

**Reply by ther3zz** [reply]
yeah, I cant get the audio to play out of my speakers.
Seems that MPD is playing back the stream, but I'm not sure how to configure things so that I can listen from a device.,

**Reply by OzGav** [reply]
You just add the radio stream URL as a manual radio station https://www.music-assistant.io/music-providers/builtin/

**Reply by gerjaneghuizen** [reply]
I use an internal link http://172.x.x.x:8000

The port is set in your Add-Ons for MPD:
![image](https://github.com/user-attachments/assets/a4e9917a-d28d-46ab-986f-cf16c92440a1)

**Reply by rrooggiieerr** [reply]
Yes you have to set _HTTPd Connection (if enabled)_, even though _httpd_output_ is enabled. To set the _HTTPd Connection_ you have to enable _Show disabled ports_

I configured http://homeassistant.local:8000/ in my LaMetric TIME to play the stream. MPD is configured to stream OGG, if your player only supports mp3, like mine does, you have to create a custom `mpd.conf` in `/share/mpd/` on your HA server.

**Comment by Silther** [comment]
I find this feature would be amazing as it would allow to use music assistant to integrate into things like the https://en.hoerbert.com/ so you could give children the option to play music in a safe environment without always having to upload things to the micro sd card.

**Comment by cpetry** [comment]
Has this been working in the past?
I don't see an option to add a track with URL.
Only playlists can be added.

The documentation says something about "tracks" section.
https://www.music-assistant.io/music-providers/builtin/

Not available on my side:
<img width="272" height="299" alt="image" src="https://github.com/user-attachments/assets/ab4108ad-aaf1-4122-ba59-f9f5ec299beb" />

**Reply by OzGav** [reply]
You dont have any tracks in your library then

**Reply by cpetry** [reply]
Correct. How can I add an URL radio then?

**Reply by OzGav** [reply]
Well you go through the radio view but this brings up an interesting point. If you don't have a radio station yet then you don't see that view. We will talk abot this internally but for now add the radio-browser provider, add a station and then you will see the view you need

**Comment by Rudd-O** [comment]
We're using this feature to stream baby audio into the bedroom.

Can we please reduce the buffering that the builtin radio provider does?

I can see there's a multi-second lag between my audio stream on my phone (or on VLC) and the audio stream on the speaker I'm playing to via MA (both a Nabu Casa VPE and an AirPlay Bose SoundTouch).

Alternatively, let's allow the user to configure this in the radio provider settings.

EDIT: let me do some more debugging, it may not be MA at fault...

EDIT2 2: yes it is.  MA is encoding to MP3 before sending to VPE (through HA via the HA provider), I guess that adds some latency, but it's multi second so it can't be that much.  I don't recall how it encodes the audio (originally PCM A-Law 8 KHz) to the SoundTouch.  Maybe I'll find a way to make the VPE player play a stream directly without intervention from Music Assistant.

EDIT 3: confirmed with `ffplay -fflags nobuffer -flags low_delay rtsp://nvr.home:8554/baby_bedroom_camera?audio=pcma` I get almost the same exact latency I get when I watch the video directly on Frigate in the same machine, but MA playback of the same stream through the VPE takes a second and a half extra to arrive to my ears, and on the Airplay Bose Soundtouch in excess of three seconds.
 

Either way, it's cool that this feature even exists.
